proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,217 @@
+While nuget.org provides a VDB, there are several reasons why customers may not be using nuget.org as a package source.
+This is especially true following the 2021 [blog post titled "Dependency confusion"](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610) which significantly increased knowledge of multi-source package substitution supply chain risks.
+Someone at Microsoft published [a whitepaper titled "3 ways to mitigate risk when using private package feeds"](https://azure.microsoft.com/mediahandler/files/resourcefiles/3-ways-to-mitigate-risk-using-private-package-feeds/3%20Ways%20to%20Mitigate%20Risk%20When%20Using%20Private%20Package%20Feeds%20-%20v1.0.pdf), so from customers points of view, it appears that Microsoft's official advice is to not use nuget.org directly, but instead use a single package source that uses nuget.org as an upstream source.
+
+While the design of NuGetAudit attempts to make it easy for upstreaming package sources to also upstream the `VulnerabilityInfo` resource, the NuGet team and customers are reliant of the feed implementor to implement the change.
+An upstreaming package source is one that either aggregates multiple package sources or caches packages from another package source.
+Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.
+
+However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
+There are also other scenarios where customers do not use nuget.org (directly) as a package source.
+
+It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
+However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### auditSources in nuget.config
+
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />`.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
```

Comment:
> When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.

This feels elegant to me. Instead of the previous text that meant "falling back to `NuGet.Configuration` for a default audit source", I like that it is either implicit that package source's vulnerability resource is used, unless `<auditSources>` is provided.

Reply:
What you're referring to was only in an "unanswered questions" section from an earlier version of the spec. That proposal was part of changing NuGetAudit to never use package sources as audit sources, and therefore we'd need another way to set up default audit sources in order to allow NuGetAudit to work by default.

It's precisely how other package ecosystems work (the tool has a hardcoded default, for both package sources, and their audit URL). NuGet supports `<clear/>` on all these "list data types", which would be a way for customers to remove the default, if we had implemented it this way instead.

My gut feeling is that using package sources as audit sources will cause more confusion than benefit, in the long term, and I don't believe that customers needing to duplicate a URL in two sections of the config file is a serious issue. But unfortunately we took an easy option with NuGetAudit V1 and now it would be a breaking change to change the behaviour to what I consider a better option.

Reply:
I'm resolving this conversation as nothing actionable for me to modify in the design spec. If you don't agree, please clarify your feedback.